linker/resolve.go

```diff
@@ -526,6 +526,13 @@ func (r *result) resolveOptionValue(handler *reporter.Handler, mc *internal.Mess
 		for _, fld := range optVal {
 			// check for extension name
 			if fld.Name.IsExtension() {
+				// Confusingly, an extension reference inside a message literal cannot refer to
+				// elements in the same enclosing message without a qualifier. Basically, we
+				// treat this as if there were no message scopes, so only the package name is
+				// used for resolving relative references. (Inconsistent protoc behavior, but
+				// likely due to how it re-uses C++ text format implementation, and normal text
+				// format doesn't expect that kind of relative reference.)
+				scopes := scopes[:1] // first scope is file, the rest are enclosing messages
```

Comment:
Verified that scopes will always have at least 1 element here: https://github.com/bufbuild/protocompile/blob/6c975192359f563aaf07d7a96676738604bca71e/linker/resolve.go#L184